python/tk_maya/panels.py

```diff
@@ -0,0 +1,106 @@
+# Copyright (c) 2013 Shotgun Software Inc.
+# 
+# CONFIDENTIAL AND PROPRIETARY
+# 
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit 
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your 
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights 
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+"""
+Panel support utilities for Maya
+"""
+import os
+import sys
+import sgtk
+from sgtk.platform.qt import QtCore, QtGui
+
+def install_callbacks(panel_id, widget_id):
+    """
+    Helper method to assist in the panel creation process.
+    This will iterate over all QT widgets and look for a panel_id
+    widget. Once found, it will install an event filter on this panel
+    to monitor its close event, so that we can gracefully handle close,
+    refresh and deallocation of the embedded tk widget when this happens.
+    
+    :param panel_id: Object name for panel
+    :param widget_id: Object name for tk widget
+    """
+    for widget in QtGui.QApplication.allWidgets():
+     if widget.objectName() == panel_id:
+         filter = CloseEventFilter(widget)
+         filter.set_associated_widget(widget_id)
+         filter.parent_closed.connect(_on_parent_closed_callback)
+         filter.parent_dirty.connect(_on_parent_refresh_callback)
+         widget.installEventFilter(filter)
+
+def _on_parent_closed_callback(widget_id):
+    """
+    Callback which fires when a panel is closed.
+    This will locate the widget with the given id
+    and close and delete this.
+    
+    :param widget_id: Object name of widget to close
+    """
+    for widget in QtGui.QApplication.allWidgets():
+        if widget.objectName() == widget_id:
+            widget.close()
+            # delete later since we are inside a slot
+            widget.deleteLater()
+    
+def _on_parent_refresh_callback(widget_id):
+    """
+    Callback which fires when a UI refresh is needed.
+    
+    :param widget_id: Object name of widget to refresh
+    """
+    for widget in QtGui.QApplication.allWidgets():
+        if widget.objectName() == widget_id:
+            # this is a pretty blunt tool, but right now I cannot
+            # come up with a better solution - it seems the internal
+            # window parenting in maya is a little off - and/or I am
+            # not parenting up the QT widgets correctly, and I think
+            # this is the reason the UI refresh isn't working correctly.
+            # the only way to ensure a fully refreshed UI is to repaint 
+            # the entire window.
+            widget.window().update()
```

Comment:
Maybe you should simply create a find_widget method and invoke it? It would avoid this pattern where you can repeatedly make this mistake?